internal/pkg/sourceinfer/analyzer.go

```diff
@@ -0,0 +1,336 @@
+contains that data, via Foo.
+
+The analysis is recursive: types identified as sources during analysis
+will be used to evaluate the sourceyness of types that depend on them.
+For example, in the below type definitions, both Qux and Quux will be
+identified as sources:
+type Qux Foo
+type Quux Qux
+`,
+	Run: run,
+	Requires: []*analysis.Analyzer{
+		inspect.Analyzer,
+		buildssa.Analyzer,
+	},
+	ResultType: reflect.TypeOf(new(ResultType)).Elem(),
+	FactTypes:  []analysis.Fact{new(inferredSource)},
+}
+
+type ObjectGraph map[types.Object][]types.Object
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	conf, err := config.ReadConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	ins := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector)
+	typeGraph := createTypeDefGraph(pass, ins)
+
+	builtSSA := pass.ResultOf[buildssa.Analyzer].(*buildssa.SSA)
+	addFieldEdges(builtSSA, typeGraph)
+
+	inferredSources := inferSources(pass, conf, typeGraph)
+
+	return inferredSources, nil
+}
+
+func createTypeDefGraph(pass *analysis.Pass, ins *inspector.Inspector) ObjectGraph {
+	defGraph := ObjectGraph{}
+
+	genDeclFilter := []ast.Node{
+		(*ast.GenDecl)(nil),
+	}
+
+	ins.Preorder(genDeclFilter, func(n ast.Node) {
+		gd, ok := n.(*ast.GenDecl)
+		if !ok {
+			return
+		}
+		for _, spec := range gd.Specs {
+			// type definition
+			ts, ok := spec.(*ast.TypeSpec)
+			if !ok {
+				continue
+			}
+
+			// type alias, e.g. type A = B
+			// we don't need to handle these, since the type will be inlined
+			if ts.Assign != token.NoPos {
+				continue
+			}
+
+			// struct definition, fields will be checked later in the analysis
+			if _, ok := ts.Type.(*ast.StructType); ok {
+				continue
+			}
+
+			typeBeingDefined := pass.TypesInfo.ObjectOf(ts.Name)
+
+			selectorFinder := &selectorFinder{nil, map[*ast.Ident]bool{}}
+			ast.Walk(selectorFinder, (ast.Node)(ts.Type))
+			for _, sel := range selectorFinder.foundSelectors {
+				named, ok := pass.TypesInfo.TypeOf(sel).(*types.Named)
+				if !ok {
+					continue
+				}
+				obj := named.Obj()
+				if len(defGraph[obj]) == 0 {
+					defGraph[obj] = make([]types.Object, 0, 1)
+				}
+				defGraph[obj] = append(defGraph[obj], typeBeingDefined)
+			}
+
+			idFinder := &identFinder{}
+			ast.Walk(idFinder, (ast.Node)(ts.Type))
+			for _, id := range idFinder.foundIdentifiers {
+				// identifier is part of a SelectorExpr and has already been handled
+				if selectorFinder.foundIdentifiers[id] {
+					continue
+				}
+				obj := pass.TypesInfo.ObjectOf(id)
+				if len(defGraph[obj]) == 0 {
+					defGraph[obj] = make([]types.Object, 0, 1)
+				}
+				defGraph[obj] = append(defGraph[obj], typeBeingDefined)
+			}
+		}
+	})
+
+	return defGraph
+}
+
+type selectorFinder struct {
+	foundSelectors   []*ast.SelectorExpr
+	foundIdentifiers map[*ast.Ident]bool
+}
+
+func (sf *selectorFinder) Visit(n ast.Node) ast.Visitor {
+	sel, ok := n.(*ast.SelectorExpr)
+	if !ok {
```

Comment:
I considered using `if sel, ok := ...; ok {`, but with the other `if` below I found it produced too much nesting.